java/ql/src/semmle/code/java/security/RequestForgery.qll

```diff
@@ -0,0 +1,330 @@
+/** Provides classes to reason about server-side request forgery (SSRF) attacks. */
+
+import java
+import semmle.code.java.frameworks.Networking
+import semmle.code.java.frameworks.ApacheHttp
+import semmle.code.java.frameworks.spring.Spring
+import semmle.code.java.frameworks.JaxWS
+import semmle.code.java.frameworks.javase.Http
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.StringFormat
+
+/**
+ * A unit class for adding additional taint steps that are specific to server-side request forgery (SSRF) attacks.
+ *
+ * Extend this class to add additional taint steps to the SSRF query.
+ */
+class RequestForgeryAdditionalTaintStep extends Unit {
+  /**
+   * Holds if the step from `pred` to `succ` should be considered a taint
+   * step for server-side request forgery.
+   */
+  abstract predicate propagatesTaint(DataFlow::Node pred, DataFlow::Node succ);
+}
+
+private class DefaultRequestForgeryAdditionalTaintStep extends RequestForgeryAdditionalTaintStep {
+  override predicate propagatesTaint(DataFlow::Node pred, DataFlow::Node succ) {
+    // propagate to a URI when its host is assigned to
+    exists(UriCreation c | c.getHostArg() = pred.asExpr() | succ.asExpr() = c)
+    or
+    // propagate to a URL when its host is assigned to
+    exists(UrlConstructorCall c | c.getHostArg() = pred.asExpr() | succ.asExpr() = c)
+    or
+    // propagate to a RequestEntity when its url is assigned to
+    exists(MethodAccess m |
+      m.getMethod().getDeclaringType() instanceof SpringRequestEntity and
+      (
+        m.getMethod().hasName(["get", "post", "head", "delete", "options", "patch", "put"]) and
+        m.getArgument(0) = pred.asExpr() and
+        m = succ.asExpr()
+        or
+        m.getMethod().hasName("method") and
+        m.getArgument(1) = pred.asExpr() and
+        m = succ.asExpr()
+      )
+    )
+    or
+    // propagate from a `RequestEntity<>$BodyBuilder` to a `RequestEntity`
+    // when the builder is tainted
+    exists(MethodAccess m, RefType t |
+      m.getMethod().getDeclaringType() = t and
+      t.hasQualifiedName("org.springframework.http", "RequestEntity<>$BodyBuilder") and
+      m.getMethod().hasName("body") and
+      m.getQualifier() = pred.asExpr() and
+      m = succ.asExpr()
+    )
+  }
+}
+
+/** A data flow sink for server-side request forgery (SSRF) vulnerabilities. */
+abstract class RequestForgerySink extends DataFlow::Node { }
+
+/**
+ * An argument to a url `openConnection` or `openStream` call
+ * taken as a sink for request forgery vulnerabilities.
+ */
+private class UrlOpen extends RequestForgerySink {
+  UrlOpen() {
```

Comment:
This class actually already exists as `UrlOpenSink` in `HttpsUrls.ql` where it's modelled with csv rows. Please share code.

Reply:
How is the set of sinks for this query supposed to relate to the set of sinks for `HttpsUrls.ql`? Superset, subset, or are they supposed to be equal? I guess @atorralba might help answer that?

Reply:
As I understand it, both sets of sinks are equal and are supposed to be equal.

In both cases you want methods that open connections, the difference is that in the SSRF query you want to find connections with user-controlled URLs, while in the "Failure to use HTTPS URLs" query you want to find connections with an unsafe protocol (HTTP), i.e. the sources are different but the sinks are exactly the same.

In fact, I think all sinks modelled here (not just `UrlOpen`) could be reused in the other query too.

Reply:
Excellent. Then this PR can easily improve the other query as a nice side-effect.